### PR TITLE
add custom medium

### PR DESCRIPTION
**Pull request by tylerflex**
Incorporate changes from #498 with the new data refactor in mind.

A few things of note:
* `CustomMedium.from_eps_raw()`allows user to construct each of the `eps_ii` data arrays from a single data array. How do we handle this on the backend? Do we want to change this behavior or add a new class method constructor, for example, for supplying eps at the center only?
* All of the medium `.eps_model()` take `frequency` argument, maybe we can change to `f` instead to be consistent with the data coordinates?

Needs backend implementation.

### CODE REVIEW

**Comment by weiliangjin2021** [comment]
A few remaining TODOs:

- [ ] How do we visualize custom medium in `plot_eps()`, or we just show the geometry, and users can visualize custom medium with PermittivityMonitor? @tylerflex 
- [ ] At the moment, I add a validator that along each dimension, at least two points are provided to enable interpolation. But we can remove this constraint: along dimensions where just one point is provided, treat it as translational-invariant by using `isel`.
- [ ] I also impose constraint on the coordinate of the supplied data that they must be monotonically increasing along each dimension to speed up interpolation. But it's not too slow when `assume_sorted=False`.

**Comment by tylerflex** [comment]
> How do we visualize custom medium in plot_eps(), or we just show the geometry, and users can visualize custom medium with PermittivityMonitor? @tylerflex
My vote is to just use the geometry and with a special handling for the medium (color) property, similar to how PEC is handled.
>  At the moment, I add a validator that along each dimension, at least two points need to be provided to enable interpolation. But we can remove this constraint: along dimensions where just one point is provided, treat it as translational-invariant by using isel.
That sounds good to me.
>  I also impose constraint on the coordinate of the supplied data that they must be monotonically increasing along each dimension to speed up interpolation. But it's not too slow when assume_sorted=False.
I think either way is fine, generally we try to have them sorted and I have no problem if we want to validate this. maybe one fast way is to call [sortby()](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.sortby.html) on each of the dims and ensure that the result is equal to the starting dataArray, ie something like `[assert da == da.sortby(dim) for dim in da.dims]`?
>  For eps_diagonal, permittivity of max{|eps|} is picked up along the space for each polarization so that autoMesh functions appropriately. But the performance of eps_diagonal doesn't look good.
What do you mean by the performance? like it is slow to compute eps_diagonal?

**Comment by weiliangjin2021** [comment]
> What do you mean by the performance? like it is slow to compute eps_diagonal?

When `n/k` is of size 200x200x1000, `eps_diagonal` and `eps_model` each will take about 5 seconds. This can slow down the process of grid generation. But I guess it is acceptable since expensive grid generation will be handled on the server.

**Comment by tylerflex** [comment]
I see. So basically we need to compute the max ε inside the geometry and use that for the mesh generation. But the current ε_diagonal returns ε at all positions and is therefore slow? maybe for ε diagonal we can just return `np.max()` of each of the ε_ii DataArrays in the supplied data, which might be faster?

**Comment by weiliangjin2021** [comment]
> But the current ε_diagonal returns ε at all positions and is therefore slow? maybe for ε diagonal we can just return np.max() of each of the ε_ii DataArrays in the supplied data, which might be faster?

That's how I handled in this PR: https://github.com/flexcompute/tidy3d/blob/60fa839c6e7ee019c4cf047bdb2a5a4dee1731bd/tidy3d/components/medium.py#L478.

**Comment by momchil-flex** [comment]
> > But the current ε_diagonal returns ε at all positions and is therefore slow? maybe for ε diagonal we can just return np.max() of each of the ε_ii DataArrays in the supplied data, which might be faster?
> 
> That's how I handled in this PR:
> 
> https://github.com/flexcompute/tidy3d/blob/60fa839c6e7ee019c4cf047bdb2a5a4dee1731bd/tidy3d/components/medium.py#L478
> 
> .

Yeah I think this makes much more sense.

> > I also impose constraint on the coordinate of the supplied data that they must be monotonically increasing along each dimension to speed up interpolation. But it's not too slow when assume_sorted=False.

> I think either way is fine, generally we try to have them sorted and I have no problem if we want to validate this. maybe one fast way is to call [sortby()](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.sortby.html) on each of the dims and ensure that the result is equal to the starting dataArray, ie something like [assert da == da.sortby(dim) for dim in da.dims]?

Maybe performance won't matter so much here, but even this sounds unnecessarily slow. Why not just something like `[assert coord == np.sort(coord) for coord in da.coords.values()]`?

> At the moment, I add a validator that along each dimension, at least two points need to be provided to enable interpolation. But we can remove this constraint: along dimensions where just one point is provided, treat it as translational-invariant by using isel.

I think easiest is `sel(.., method="nearest")`. That's what I do for the custom source to get the E and H field values which are slightly offset in the normal direction on the Yee grid.

**Comment by weiliangjin2021** [comment]
Now `sim.epsilon` for "Ex, Ey, Ez" components are working, I'll push another commit so that it works for "centers" and "boundaries". One question for @tylerflex: should we add a `plot_eps_on_grid` method so that the users can at least view the epsilon profile before running the local mode solver?

**Comment by flexcompute-readthedocs** [comment]
sounds good to me

**Comment by weiliangjin2021** [comment]
Regarding visualization, `plot` will not distinguish between structures made of regular medium and custom medium, while `plot_eps` will do. See an example here: The background medium and the medium for the cylinder are custom, while the box is made of regular medium.
plot
plot_eps

File: tidy3d/components/medium.py [hunk 1]
```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
...
+                val.field_components[comp], val.field_components[comp].f
+            )
+            if np.any(eps_inf.values < 1):
+                raise SetupError(
+                    "Permittivity at infinite frequency at any spatial point "
+                    "must be no less than one."
+                )
+            if np.any(sigma.values < 0):
+                raise SetupError(
+                    "Negative imaginary part of refrative index leads to a gain medium."
```
Comment (tylerflex) [inline tidy3d/components/medium.py:402]: maybe add ", which is not supported."

File: tidy3d/components/medium.py [hunk 2]
```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
...
-        eps_zz = self.zz.eps_model(frequency)
-        return np.mean((eps_xx, eps_yy, eps_zz))
+        Parameters
+        ----------
+        frequency : float
+            Frequency to evaluate permittivity at (Hz).
+
+        Returns
+        -------
+        PermittivityDataset
```
Comment (tylerflex) [inline tidy3d/components/medium.py:418]: :class: syntax here

File: tidy3d/components/medium.py [hunk 3]
```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
...
     def eps_diagonal(self, frequency: float) -> Tuple[complex, complex, complex]:
-        """Main diagonal of the complex-valued permittivity tensor as a function of frequency."""
+        """Main diagonal of the complex-valued permittivity tensor
+        at ``frequency``. Spatially, we take max{|eps|}, so that autoMesh generation
+        works appropriately.
+        """
+        eps_freq = self.eps_dataset_freq(frequency)
+        eps_np_list = [
+            np.array(eps_freq.field_components[comp]).ravel()
+            for comp in ["eps_xx", "eps_yy", "eps_zz"]
```
Comment (tylerflex) [inline tidy3d/components/medium.py:473]: what happens if any of these fields are missing? Also, might be a bit more simple to do
```py
eps_np_list = [np.array(sclr_fld).ravel() for _, sclr_fld in eps_freq.field_components.items()]
```

Comment (tylerflex) [reply tidy3d/components/medium.py:473]: same general comments at line ~460 above

File: tidy3d/components/medium.py [hunk 4]
```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
...
-        eps_yy = self.yy.eps_model(frequency)
-        eps_zz = self.zz.eps_model(frequency)
-        return (eps_xx, eps_yy, eps_zz)
+    @ensure_freq_in_range
+    def eps_model(self, frequency: float) -> complex:
+        """Spatial and poloarizaiton average of complex-valued permittivity
+        as a function of frequency.
+        """
+        eps_freq = self.eps_dataset_freq(frequency)
+        eps_arrays = [eps_freq.field_components[comp] for comp in ["eps_xx", "eps_yy", "eps_zz"]]
```
Comment (tylerflex) [inline tidy3d/components/medium.py:484]: can probably just use
```py
eps_arrays = list(eps_freq.field_components.values()]
```

Comment (tylerflex) [reply tidy3d/components/medium.py:484]: or just skip this step and do
```py
eps_array_avgs = [np.mean(eps_array) for _, eps_array in eps_freq.field_components.items()]
```

File: tidy3d/components/medium.py [hunk 5]
```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
...
+        return np.mean(eps_array_avgs)
 
-    @add_ax_if_none
-    def plot(self, freqs: float, ax: Ax = None) -> Ax:
-        """Plot n, k of a :class:`Medium` as a function of frequency."""
+    @classmethod
+    def from_eps_raw(
+        cls, eps: ScalarFieldDataArray, interp_method: InterpMethod = "nearest"
+    ) -> CustomMedium:
+        """Construct a :class:`CustomMedium` from datasets containing raw permittivity values.
```
Comment (tylerflex) [inline tidy3d/components/medium.py:492]: put `.` in front of `CustomMedium`?

File: tidy3d/components/medium.py [hunk 6]
```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
...
 
-        freqs = np.array(freqs)
-        freqs_thz = freqs / 1e12
+        Parameters
+        ----------
+        eps : :class:`.ScalarFieldDataArray`
+            Dataset containing complex-valued permittivity as a function of space.
+        interp_method : InterpMethod, optional
```
Comment (tylerflex) [inline tidy3d/components/medium.py:498]: might want to provide the raw literal values accepted here, or :class: link to InterpMethod if that works.

File: tidy3d/components/medium.py [hunk 7]
```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
...
-        ax.legend()
-        ax.set_aspect("auto")
-        return ax
+        Parameters
+        ----------
+        n : :class:`.ScalarFieldDataArray`
+            Real part of refractive index.
+        k : :class:`.ScalarFieldDataArray` = None
+            Imaginary part of refrative index.
+        interp_method : InterpMethod, optional
```
Comment (tylerflex) [inline tidy3d/components/medium.py:526]: same comment about literal values or linking to this type if possible.

File: tidy3d/components/medium.py [hunk 8]
```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
...
+                Interpolation method to obtain permittivity values that are not supplied
+                at the Yee grids.
+
+        Returns
+        -------
+        :class:`CustomMedium`
```
Comment (tylerflex) [inline tidy3d/components/medium.py:532]: dot in front

File: tidy3d/components/medium.py [hunk 9]
```diff
@@ -756,8 +993,94 @@ def pole_residue(self):
         )
 
 
+IsotropicMediumType = Union[Medium, PoleResidue, Sellmeier, Lorentz, Debye, Drude]
+
+
+class AnisotropicMedium(AbstractMedium):
```
Comment (tylerflex) [inline tidy3d/components/medium.py:999]: Not sure why this change is here, maybe it needs to be rebased against pre/1.9.0?

**Review by tylerflex — COMMENTED** [review]
Looks good! Just some formatting and minor improvements from me.

File: tidy3d/components/medium.py (hunk 3 above)
Comment (weiliangjin2021) [reply tidy3d/components/medium.py:473]: One question here: when unpacked with `.items`, will it be in order (eps_xx, eps_yy, eps__zz)?

Comment (tylerflex) [reply tidy3d/components/medium.py:473]: Should be, yes. Because after python 3.6 or something dictionaries are ordered.

File: tidy3d/components/medium.py (hunk 6 above)
Comment (weiliangjin2021) [reply tidy3d/components/medium.py:498]: As for the raw literal values, is it like this:
`interp_method : ["nearest", "linear"], optional`

File: tidy3d/components/medium.py (hunk 3 above)
Comment (weiliangjin2021) [reply tidy3d/components/medium.py:473]: > what happens if any of these fields are missing? 

All the fields are required as defined in `PermittivityDataset`: https://github.com/flexcompute/tidy3d/blob/28d74cfc9965fe3aea79a3198fe558c6e2561e47/tidy3d/components/data/dataset.py#L369-L383

Comment (tylerflex) [reply tidy3d/components/medium.py:473]: ah ok.

**Comment by weiliangjin2021** [comment]
Commits are squashed. Planning to merge to pre/1.9.0 this afternoon. @tylerflex
